File: javascript/beer-song/beer-song.js
class BeerSong {
  verse(num) {
    const verse = this.template(
      this.bottles(num),
      this.lineThree(num),
      this.bottles(num - 1)
    );

    return verse.charAt(0).toUpperCase() + verse.slice(1);
  }

  sing(start, end = 0) {
    let song = '';
    let first = start;

    for (first; first > end - 1; first -= 1) {
      song += this.verse(first);
      if (first !== end) {
        song += '\n';
      }
    }
    return song;
  }

  bottles(num) {
    if (num < 0) return '99 bottles of beer';
    if (num > 1) return `${num} bottles of beer`;
    if (num === 1) return '1 bottle of beer';
    return 'no more bottles of beer';
  }

  lineThree(num) {
    if (num > 1) return 'Take one down and pass it around';
    if (num === 1) return 'Take it down and pass it around';
    return 'Go to the store and buy some more';
  }

  template(bottles, lineThree, bottlesTwo) {
    return `${bottles} on the wall, ${bottles}.\n` +
           `${lineThree}, ${bottlesTwo} on the wall.\n`;
  }
}

module.exports = BeerSong;

File: ruby/beer-song/beer_song.rb
module BookKeeping
  VERSION = 3
end

class BeerSong
  def verse(num)
    verse = template(
      bottles(num),
      line_three(num),
      bottles(num - 1)
    )
    verse[0].upcase + verse.slice(1, verse.length)
  end

  def verses(start, finish = 0)
    song = ''

    start.downto(finish) do |i|
      song << verse(i)
      song << "\n" if i != finish
    end

    song
  end

  def bottles(num)
    return '99 bottles of beer' if num < 0
    return "#{num} bottles of beer" if num > 1
    return '1 bottle of beer' if num == 1
    'no more bottles of beer'
  end

  def line_three(num)
    count = num > 1 ? 'one' : 'it'
    return "Take #{count} down and pass it around" if num > 0
    'Go to the store and buy some more'
  end

  def template(bottle_count, line_three, bottle_count_two)
    "#{bottle_count} on the wall, #{bottle_count}.\n" \
      "#{line_three}, #{bottle_count_two} on the wall.\n"
  end
end
